### PR TITLE
Adding auth0 OAuthenticator implementation

### PR DESCRIPTION
**Pull request by namdets**
Auth0 is a OAuth2 authentication aggregator service that supports mixing LDAP, SAML, OAuth2, etc. backing authentication providers.

We find it very useful.

You can test it by creating a free account here : https://auth0.com/opensource

### CODE REVIEW

File: oauthenticator/auth0.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+    oauth_variable_config = Any(
+        help="""Any callable that returns a dictionary of oauth variables by the following names:
+        OAUTH_CLIENT_ID
+        OAUTH_CLIENT_SECRET
+        OAUTH_CALLBACK_URI
+        """
+    ).tag(config=True)
+
+    def oauth_environment_variable_config(self):
```
Comment (minrk) [inline oauthenticator/auth0.py]: Can we skip this, and leave the getting of CLIENT_ID, etc. from the env the same as all the other implementations?

**Comment by namdets** [comment]
Wow, Thanks for that!

So much cleaner!

**Comment by namdets** [comment]
Hold on, there may now be a bug with the environment variable based config. Update incoming.

**Comment by namdets** [comment]
OK, false alarm. I had a bug in my environment config, not the code.

I'm happy if you are.

**Comment by minrk** [comment]
Thanks!

**Pull request merged**